ESQL: Enable arithmetics for durations and periods

alex-spies opened the pull request:
This partially solves #99293.

- Enable date math expressions like `(1year + 1month) + now()` and `(1hour + 1second) + now()`, where the durations/periods are added before being added to a datetime. This extends to expressions like `1year + month + now()` (without parentheses).
- To achieve this, enable folding of duration/period literals.
- Provide user-friendly error messages for date time arithmetic overflows, and for illegal expressions like `eval x = 1week + 1month` whose type is not representable.

What's not covered from #99293: Mixed duration/period expressions like `(1year + 1hour) + now()`, resp. `1year + 1hour + now()`.

alex-spies committed:
DRAFT: allow adding/subtracting date durations

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/date.csv-spec:
```diff
@@ -614,3 +614,17 @@ birth_date:datetime
 ;
+
+dateMathAddDatePeriodsThenAddToDatetime
+row n = to_dt("1953-04-20T00:00:00.000Z") | eval then = (1 year + 1 month + 1 day) + n | keep then;
+
+then:datetime
+1954-05-21T00:00:00.000Z
+;
+
+dateMathSubtractDatePeriodsThenAddToDatetime
+row n = to_dt("1954-05-21T00:00:00.000Z") | eval then = (-1 year - 1 month - 1 day) + n | keep then;
```
These cases already pass fine.

Implementing date math for cases where we add/subtract time periods (hours, minutes etc.) can be done easily in an analogous way.

Implementing mixed expressions (date and time periods, e.g. `1 year + 1 hour`) are possible but require at least a follow-up PR.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java:
```diff
@@ -60,6 +62,7 @@ public final class EsqlDataTypes {
         DATETIME,
         DATE_PERIOD,
         TIME_DURATION,
+        DURATION,
```
Adding `DURATION` is not necessary for this PR, but an outlook for how the functionality could look like when we allow combined expressions like `1year + 1 hour`.

bpintea replied:
I guess this would make sense to add only when properly/completely introducing this type and the allowed operations.
But not convinced yet this is absolutely needed; we could support something like `| EVAL now() + (1 year + 1 second) - (3 minutes + 1 month)` if needed, but even then we might be able to drop the parentheses (unless we plan to support multiplications, like for SQL's intervals) and rewrite the expression into `| EVAL now() + 1 year - 1 month + 1 second - 3 minutes` which wouldn't require an extra type.

alex-spies replied:
Yep, this needs extra consideration. I will definitely take it out of this PR.

I agree that this could be solved without introducing an additional type by rewriting the expression.

alex-spies committed:
Apply spotless

alex-spies committed:
Improve tests

alex-spies committed:
Add tests for time durations

alex-spies committed:
Implement arithmetics for time periods

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/date.csv-spec:
```diff
@@ -614,3 +614,31 @@ birth_date:datetime
 1953-04-21T00:00:00.000Z
 ;
+
+dateMathAddDatePeriodsThenAddToDatetime
+row n = to_dt("1953-04-04T00:00:00.000Z") | eval then = (4 years + 3 months + 2 weeks + 1 day) + n | keep then;
+
```
Please add some tests without parenthesis just to be sure.

bpintea replied:
I guess the parentheses are needed to trigger the type-specific folding. There are tests without parentheses already.

alex-spies replied:
The parentheses are actually redundant. I just wanted tests that make it obvious that the expression adds the durations/periods _before_ adding to a datetime, which is the currently unsupported case.

alex-spies replied:
Added more test cases! Also added tests for subtractions from datetimes combined with first folding period/duration constants.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java:
```diff
@@ -50,16 +54,49 @@ protected TypeResolution resolveType() {
-                    format(null, "[{}] has arguments with incompatible types [{}] and [{}]", symbol(), leftType, rightType)
-                );
+            if (argumentOfType(DataTypes::isDateTime) != null && argumentOfType(EsqlDataTypes::isTemporalAmount) != null) {
+                return TypeResolution.TYPE_RESOLVED;
             }
-            return TypeResolution.TYPE_RESOLVED;
+            if (EsqlDataTypes.isTemporalAmount(leftType) && EsqlDataTypes.isTemporalAmount(rightType)) {
+                return TypeResolution.TYPE_RESOLVED;
+            }
```
These two conditions look redundant - isn't argumentOfType already checking the TemporalAmount?

alex-spies replied:
The first condition checks whether there is one datetime and one duration argument (where duration does not include datetime) in any order. The second checks whether both arguments are durations, so the two conditions are mutually exclusive.

I'll rewrite this in a way that is hopefully clearer!

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java:
```diff
@@ -50,16 +54,49 @@ protected TypeResolution resolveType() {
         return super.resolveType();
     }
 
+    @Override
+    public final Object fold() {
+        DataType leftDataType = left().dataType();
+        DataType rightDataType = right().dataType();
+        // TODO: handle time_durations
+        // TODO: handle mixed time_durations and date_periods
+        if (leftDataType == DATE_PERIOD && rightDataType == DATE_PERIOD) {
+            // Both left and right expressions are temporal amounts; we can assume they are both foldable.
```
This needs to be extracted and be applicable in block friendly fashion.
Consider `eval m = 1 minute | eval y = 1 year | eval z = m + y` - the code should be in one place regardless if it's a constant (ConstantBlock) or actual backing block.

bpintea replied:
These types can only make sense in an expression involving a DATETIME input and a DATETIME output, since we only have blocks(/vectors) of this type.

alex-spies replied:
Yep, what @bpintea said that was my thinking as well. Date math is a bit special as we never want to create blocks of an other type than `DATETIME`. So, the only time when we want to have arithmetics between date periods or time durations is when folding constants.

@costin, I can probably extract this into block friendly code, but I think that will create additional complexity that might not be warranted. Currently, the evaluators for date math are based on the fact that a date time block will be added to a constant value (see[ evaluator definition here](https://github.com/elastic/elasticsearch/blob/aff49dc69a489a40f2aa4a2ae8c31b3c4825b2a8/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/Add.java#L87-L89)); so, making the date math logic block based will likely require bigger changes.

If that's okay with you, @costin , I'd prefer not to block-ify the duration/period arithmetics. I think I can put the code for the actual arithmetics into the corresponding operator class, though (`Add.java` and `Sub.java`), then at least all processing code is nicely in one place, even if the code for processing periods/durations does not generate an evaluator.

costin reviewed (changes requested):
Left a round of comments

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/Sub.java:
```diff
@@ -44,7 +44,7 @@ public Sub(Source source, Expression left, Expression right) {
     protected TypeResolution resolveType() {
         TypeResolution resolution = super.resolveType();
-        if (resolution.resolved() && EsqlDataTypes.isDateTimeOrTemporal(dataType()) && DataTypes.isDateTime(left().dataType()) == false) {
+        if (resolution.resolved() && EsqlDataTypes.isDateTimeOrTemporal(dataType()) && DataTypes.isDateTime(right().dataType())) {
```
The changed format works because `super.resolveType()` ensures that they the operands are of different type, but the original form makes it more clear about the LHS type requirement.

alex-spies replied:
Hm, to allow folding of durations/periods with SUB, I need to allow the left hand side to be something different than datetime. In this respect, the two conditions are not equivalent and I need this change to allow e.g. `1week - 1day` while still disallowing `1week - now()`. This is the most concise way I came up with, and it seems logical to me: it says "you cannot subtract a datetime from something else".

I'll add a comment to make sure this is as clear as can be; if there's something else I can do to make things clear, let me know!

bpintea reviewed (commented):
IMO the folding approach here is good, but ideally we would extract/generalise this approach for other math operations as well, in the sense noted here: https://github.com/elastic/elasticsearch/issues/99502

alex-spies committed:
Generalize tests, add tests without parentheses

alex-spies committed:
Revise type resolution

alex-spies committed:
Add comment to clarify SUB type resolution

alex-spies committed:
Remove datatype for combined durations and periods

alex-spies committed:
Move actual constant folding logic into Add/Sub

alex-spies committed:
Fix typos

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-add-subtract-durations

alex-spies committed:
Apply spotless

alex-spies committed:
Increase range of random test periods/durations

alex-spies committed:
Add unit tests for duration/period arithmetics

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java:
```diff
@@ -48,30 +51,85 @@ interface DatetimeArithmeticEvaluator extends TriFunction<Source, ExpressionEval
+    protected Duration processTimeDurations(Duration left, Duration right) {
+        throw new UnsupportedOperationException("processing of date durations is unsupported for [" + op().symbol() + "]");
+    }
+
+    @Override
+    public final Object fold() {
+        DataType leftDataType = left().dataType();
+        DataType rightDataType = right().dataType();
+        if (leftDataType == DATE_PERIOD && rightDataType == DATE_PERIOD) {
+            // Both left and right expressions are temporal amounts; we can assume they are both foldable.
+            Period l = (Period) left().fold();
+            Period r = (Period) right().fold();
+            return processDatePeriods(l, r);
+        }
+        if (leftDataType == TIME_DURATION && rightDataType == TIME_DURATION) {
+            // Both left and right expressions are temporal amounts; we can assume they are both foldable.
+            Duration l = (Duration) left().fold();
+            Duration r = (Duration) right().fold();
+            return processTimeDurations(l, r);
+        }
+        return super.fold();
+    }
+
     @Override
     public Supplier<ExpressionEvaluator> toEvaluator(Function<Expression, Supplier<ExpressionEvaluator>> toEvaluator) {
         return dataType() == DataTypes.DATETIME
             ? () -> datetimes.apply(
                 source(),
-                toEvaluator.apply(argumentOfType(DataTypes::isDateTime)).get(),
-                (TemporalAmount) argumentOfType(EsqlDataTypes::isTemporalAmount).fold()
+                toEvaluator.apply(argumentWhich(DataTypes::isDateTime)).get(),
+                (TemporalAmount) argumentWhich(EsqlDataTypes::isTemporalAmount).fold()
             )
             : super.toEvaluator(toEvaluator);
     }
 
-    private Expression argumentOfType(Predicate<DataType> filter) {
+    private Expression argumentWhich(Predicate<DataType> filter) {
         return filter.test(left().dataType()) ? left() : filter.test(right().dataType()) ? right() : null;
     }
```
There are two arguments - if one is DateTime, the other has to be TemporalAmount, otherwise flip them.
Easier then doing 4 method invocations of a method that has two nested method invocations internally.

There's already a type check in resolve() where the explicit types are determined.

alex-spies replied:
I'll simplify `toEvaluator` and `resolveType` to reduce the necessary number of method invocations.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java:
```diff
@@ -48,30 +51,85 @@ interface DatetimeArithmeticEvaluator extends TriFunction<Source, ExpressionEval
+     * @param right the right period
+     * @return the result of the evaluation
+     */
+    protected Period processDatePeriods(Period left, Period right) {
```
DateArithmetic()/PeriodArithmetic() or simply arithmetic(Period left, Period right)

alex-spies replied:
Sounds good, although `processPeriods` may be more in line with `processDatetimes`, `processInts` etc. To still have a verb, though, I can make this e.g. `computeDateArithmetic`.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java:
```diff
@@ -48,30 +51,85 @@ interface DatetimeArithmeticEvaluator extends TriFunction<Source, ExpressionEval
+     * @param right the right duration
+     * @return the result of the evaluation
+     */
+    protected Duration processTimeDurations(Duration left, Duration right) {
```
TimeArithmetic/DurationArithmetic or simply arithmetic(Duration left, Duration right).

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/date.csv-spec:
```diff
@@ -614,3 +614,60 @@ birth_date:datetime
 ;
+
+dateMathAddDatePeriodsFromLeft
+row n = to_dt("1953-04-04T00:00:00.000Z") | eval then = 4 years + 3 months + 2 weeks + 1 day + n | keep then;
```
Add also 1 year + 1 year  + 2 years, 1 months + 1 week + 1 month+ + 1 week , etc...
Mix the types to make sure they properly get combined.

alex-spies replied:
The unit tests should already cover a lot of ground there, but I'll happily add CSV tests as well.

costin reviewed (commented):
Left another round

alex-spies committed:
Verify output type of EVAL is representable

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-add-subtract-durations

alex-spies committed:
Handle arithmetic exceptions while folding

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java:
```diff
@@ -9,21 +9,37 @@
 
 abstract class DateTimeArithmeticOperation extends EsqlArithmeticOperation {
 
+    private class IllegalTemporalValueException extends EsqlClientException {
```
This still needs a bit of polish, at least doc strings that explain why this exception class is even needed.

(In case of overflows during folding, the user should see a nice error message and we shouldn't treat this as 500 but 400.)

alex-spies reviewed (commented):
I'll address the remarks (thanks!), will add tests for overflows when folding, then this should be ready for review.

alex-spies committed:
Improve IllegalTemporalValueException

alex-spies committed:
Add test for overflows in literal date math

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/date.csv-spec:
```diff
@@ -614,3 +614,60 @@ birth_date:datetime
+;
+
+dateMathSubtractDatePeriodsFromLeft
+row n = to_dt("1957-07-19T00:00:00.000Z") | eval then = -4 years - 3 months - 2 weeks - 1 day + n | keep then;
```
The second line will not work exactly like this because `z = -2 weeks - 1 day` tries to assign an unrepresentable value; I did `then = y + (-2 weeks -1 day)` instead.

costin replied:
> The second line will not work exactly like this because z = -2 weeks - 1 day
That calls for another test that handles negation of date intervals.

alex-spies replied:
That is a very good remark. I thought negations were already handled and tested fine (there's multiple tests which contain expressions like `-1hour` etc.), but it turns out that negation fails when used over a sum like `-(1hour + 2hour)`.

I'll take care of that.

alex-spies replied:
So, handling expressions like `now() + (-(1hour + 2hour))` will require increasing the scope of this endeavor to generalize folding of `Neg` as well, together with associated unit and csv tests. I'd like to do this in a follow-up PR because that for sure warrants a separate review.

alex-spies committed:
Add test that propagates date math through evals

alex-spies committed:
Clean up csv test names and order

alex-spies committed:
processDatePeriods, processTimeDurations -> public

alex-spies committed:
Simplify DateTimeArithmeticOperation methods

Simplify type resolution and toEvaluator methods; remove
`argumentOfType`/`argumentWhich` methods.

alex-spies committed:
Rename processDatePeriods/processTimeDurations

Use computeDateArithmetic and computeTimeArithmetic, which is closer to
the suggestions from review.

alex-spies committed:
More csv tests

alex-spies committed:
Update docs/changelog/99432.yaml

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-add-subtract-durations

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

alex-spies commented:
Alright, this is ready for review now!

alex-spies committed:
Update changelog

Mention issue number.

alex-spies committed:
Improve error message

alex-spies committed:
Remove redundant accessor for op

EsqlArithmeticOperation does not need to expose its op symbol anymore.

alex-spies committed:
Fix verifier test

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-add-subtract-durations

alex-spies committed:
Update docs/changelog/99432.yaml

elasticsearchmachine commented:
Hi @alex-spies, I've updated the changelog YAML for you.

alex-spies committed:
Fix changelog

alex-spies committed:
Fix REST tests

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java:
```diff
@@ -55,31 +79,100 @@ ExpressionEvaluator apply(
+     * @param right the right period
+     * @return the result of the evaluation
+     */
+    abstract Period computeDateArithmetic(Period left, Period right);
```
nit: could be just `fold(Period, Period)` (and corresponding `fold(Duration, Duration` below) to avoid the "full" date confusion.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -236,6 +238,19 @@ private static Collection<Failure> validateRow(Row row) {
+            DataType dataType = field.dataType();
+            if (EsqlDataTypes.isRepresentable(dataType) == false) {
+                failures.add(
+                    fail(field, "EVAL does not support type [{}] in expression [{}]", dataType.typeName(), field.child().sourceText())
```
Wondering if for numeric types we shouldn't suggest a conversion, à la: `cannot use [{}]  of type [{}] in an EVAL assignment. Convert to a wider type`.

alex-spies replied:
Hm, this would only apply to some of the [unrepresentable data types](https://github.com/elastic/elasticsearch/blob/c663c831b7e334f980667686d1b0555a605f4bdf/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java#L151). Converting to a wider type does not really apply to a date period, for instance, right?

But, we could add a couple more type specific error messages, like "convert to a wider type" for short, half float etc. and "can use period/duration only in combination with a datetime" or so for periods/durations.

I planned to do another small PR once this is merged to have good error messages for cases when users add durations and periods (which will inevitably happen IMO until we implement mixed durations and periods). I can add type-specific verifier errors for EVAL there :)

bpintea replied:
> Hm, this would only apply to some of the [unrepresentable data types](https://github.com/elastic/elasticsearch/blob/c663c831b7e334f980667686d1b0555a605f4bdf/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java#L151). Converting to a wider type does not really apply to a date period, for instance, right?

Right, just for numeric types (`DataType#isNumeric()`).

> when users add durations and periods

Right, we can suggest reordering there.

bpintea reviewed (approved):
Lgtm, thanks. I think it's a step forwad and we'll need the new time span folding also for arbitrary mix of durations and periods, if we'll want to support that later.

alex-spies committed:
Rename compute...Arithmetic simply to fold

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java:
```diff
@@ -9,21 +9,45 @@
+    /**
+     * Custom exception to handle e.g. overflows when folding temporal values; we want to set the correct HTTP status (400).
+     */
+    private static class IllegalTemporalValueException extends EsqlClientException {
+        protected IllegalTemporalValueException(String message, Object... args) {
+            super(message, args);
+        }
+
+        @Override
+        public RestStatus status() {
+            return RestStatus.BAD_REQUEST;
+        }
+
+        public static IllegalTemporalValueException fromArithmeticException(Source source, ArithmeticException e) {
+            return new IllegalTemporalValueException("arithmetic exception in expression [{}]: [{}]", source.text(), e.getMessage());
+        }
+    }
```
An exception has 'cognitive' value - it should be generic enough that enough places encounter the same condition that an exception makes sense to be created.

This is rarely the case (which is why *QL in general has only a couple of exceptions) -  just return an EsqlClientException with the correct message.

alex-spies replied:
I tried just using EsqlClientException before I arrived at this solution; the problem with `EsqlClientException` is that it is abstract, so I can only create an anonymous class that inherits from it.

This works, but the user facing error message that it creates when an overflow is triggered during folding is not so great:
```
{"type":"",
  "reason":"arithmetic exception in expression [2147483647 day + 1 day]: [integer overflow]"}
```
Notice the empty type string. Since the point of this exception is providing a nicer user-facing error, I don't think an anonymous class is a good solution.

With the current solution, the error message looks like this instead:
```
{"type":"illegal_temporal_value_exception",
  "reason":"arithmetic exception in expression [9223372036854775807 second + 1000 millisecond]: [long overflow]"}
```

I'll add this to the javadoc of the class to make it clear.

(I also contemplated either adding a more generic named exception, like `FoldingException` or so - but the case that it's used in only 2 places in just 1 class made that feel very awkward. The other option would be to make EsqlClientException just not abstract, but that again seemed like being against EsqlClientException's intended purpose. If you'd like me to go with any of these solutions, instead, just let me know and I'll create a follow-up PR!

There's also a solid chance that there may be more situations where a user-caused error will happen during folding time, at which point we can also factor out `IllegalTemporalValueException` into a more generic exception class.)

alex-spies replied:
As it turns out, I need an exception to communicate arithmetic exceptions when folding `Neg` as well. So, the best is probably to factor this out into a dedicated `DateMathException`.

I'll make a follow-up PR to handle `Neg` and will do that there.

costin reviewed (approved):
Left another round of comments - after addressing them the ticket is good to go.
